/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax="proto2";

option java_multiple_files = true;
option java_package = "org.apache.wayang.commons.serializable";

/*Java Operator or concatenated Python Pipeline*/
message OperatorProto{
  optional string id = 1;
  optional string type = 2;
  optional bytes udf = 3;
  optional string path = 4;
  // repeated OperatorProto predecessors = 5;
  repeated string predecessors = 5;
  // repeated OperatorProto successors = 6;
  repeated string successors = 6;
  map<string, string> parameters = 7;
}

message PlanProto{

  enum TypesProto{
    string = 0;
    integer = 1;
  }

  repeated OperatorProto sources = 1;
  repeated OperatorProto operators = 2;
  repeated OperatorProto sinks = 3;
  optional TypesProto input = 4;
  optional TypesProto output = 5;

}

message ContextProto{

  enum PlatformProto{
    java = 0;
    spark = 1;
  }

  repeated PlatformProto platforms = 1;
}

message WayangPlanProto{

  optional PlanProto plan = 1;
  optional ContextProto context = 2;

}